File: android/packages/components/src/main/kotlin/io/ente/components/Chips.kt
package io.ente.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.selection.toggleable
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Icon
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp

@Composable
public fun FilterChip(
    label: String,
    selected: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
    leading: @Composable (() -> Unit)? = null,
    trailing: @Composable (() -> Unit)? = null,
) {
    val palette = LocalEntePalette.current
    val background =
        if (selected) {
            if (palette.isDark) Color(0xFFF4F4F4) else Color(0xFF161616)
        } else {
            palette.surface
        }
    val foreground =
        when {
            selected -> palette.reverseText
            enabled -> palette.mutedText
            else -> palette.disabledText
        }
    val startPadding =
        when {
            leading != null && trailing == null -> EnteSpacing.md
            trailing != null -> EnteSpacing.lg
            else -> 18.dp
        }
    val endPadding =
        when {
            leading != null && trailing == null -> if (selected) EnteSpacing.md else EnteSpacing.lg
            trailing != null || selected -> EnteSpacing.md
            else -> 18.dp
        }

    Surface(
        color = background,
        contentColor = foreground,
        shape = RoundedCornerShape(percent = 50),
        modifier =
            modifier.heightIn(min = 40.dp).toggleable(
                value = selected,
                enabled = enabled,
                role = Role.Button,
            ) {
                onClick()
            },
    ) {
        Row(
            modifier =
                Modifier.padding(
                    start = startPadding,
                    end = endPadding,
                    top = EnteSpacing.md,
                    bottom = EnteSpacing.md,
                ),
            horizontalArrangement = Arrangement.Center,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            leading?.let {
                Box(Modifier.size(16.dp), contentAlignment = Alignment.Center) { it() }
                Spacer(Modifier.width(EnteSpacing.sm))
            }
            Text(label, maxLines = 1, overflow = TextOverflow.Ellipsis, style = EnteTypography.mini)
            if (trailing != null || selected) {
                Spacer(Modifier.width(EnteSpacing.sm))
                if (trailing != null) {
                    Box(Modifier.size(16.dp), contentAlignment = Alignment.Center) { trailing() }
                } else {
                    Icon(
                        painterResource(R.drawable.component_close),
                        null,
                        Modifier.size(14.dp),
                        tint = foreground,
                    )
                }
            }
        }
    }
}

@Composable
public fun Tag(
    label: String,
    selected: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
    leading: @Composable (() -> Unit)? = null,
    trailing: @Composable (() -> Unit)? = null,
) {
    val palette = LocalEntePalette.current
    Surface(
        color = if (selected) palette.primary else palette.surface,
        contentColor =
            if (selected) Color.White else if (enabled) palette.mutedText else palette.disabledText,
        shape = RoundedCornerShape(EnteRadius.large),
        modifier =
            modifier.heightIn(min = 44.dp).toggleable(
                value = selected,
                enabled = enabled,
                role = Role.Button,
            ) {
                onClick()
            },
    ) {
        Row(
            modifier =
                Modifier.padding(
                    start = if (leading != null) EnteSpacing.md else EnteSpacing.xl,
                    end = if (trailing != null) EnteSpacing.md else EnteSpacing.xl,
                    top = EnteSpacing.md,
                    bottom = EnteSpacing.md,
                ),
            horizontalArrangement = Arrangement.Center,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            leading?.let {
                Box(Modifier.size(20.dp), contentAlignment = Alignment.Center) { it() }
                Spacer(Modifier.width(EnteSpacing.xs))
            }
            Text(label, maxLines = 1, overflow = TextOverflow.Ellipsis, style = EnteTypography.body)
            trailing?.let {
                Spacer(Modifier.width(EnteSpacing.xs))
                Box(Modifier.size(20.dp), contentAlignment = Alignment.Center) { it() }
            }
        }
    }
}

File: apple/packages/EnteComponents/Sources/EnteComponents/Chips.swift
import SwiftUI

public struct FilterChip: View {
    @Environment(\.entePalette) private var palette
    @Environment(\.isEnabled) private var enabled
    private let label: String
    private let image: Image?
    private let selected: Bool
    private let action: () -> Void

    public init(
        _ label: String,
        image: Image? = nil,
        selected: Bool,
        action: @escaping () -> Void,
    ) {
        self.label = label
        self.image = image
        self.selected = selected
        self.action = action
    }

    public var body: some View {
        Button(action: action) {
            HStack(spacing: EnteSpacing.sm) {
                if let image {
                    image.resizable().renderingMode(.template).scaledToFit()
                        .frame(width: 16, height: 16)
                        .accessibilityHidden(true)
                }
                Text(label)
                if selected {
                    Glyph.close.image.resizable().scaledToFit()
                        .frame(width: 14, height: 14)
                        .accessibilityHidden(true)
                }
            }
            .font(EnteTypography.mini)
            .lineLimit(1)
            .padding(.leading, image == nil ? 18 : EnteSpacing.md)
            .padding(
                .trailing,
                selected ? EnteSpacing.md : (image == nil ? 18 : EnteSpacing.lg),
            )
            .padding(.vertical, EnteSpacing.md)
            .frame(minHeight: 40)
        }
        .buttonStyle(.plain)
        .foregroundStyle(foreground)
        .background(background)
        .clipShape(Capsule())
        .accessibilityAddTraits(selected ? .isSelected : [])
    }

    private var background: Color {
        selected ? (palette.isDark ? Color(hex: 0xF4F4F4) : Color(hex: 0x161616)) : palette.surface
    }

    private var foreground: Color {
        selected ? palette.reverseText : (enabled ? palette.mutedText : palette.disabledText)
    }
}

public struct Tag: View {
    @Environment(\.entePalette) private var palette
    @Environment(\.isEnabled) private var enabled
    private let label: String
    private let selected: Bool
    private let action: () -> Void

    public init(
        _ label: String,
        selected: Bool,
        action: @escaping () -> Void,
    ) {
        self.label = label
        self.selected = selected
        self.action = action
    }

    public var body: some View {
        Button(action: action) {
            Text(label)
                .font(EnteTypography.body)
                .lineLimit(1)
                .padding(.horizontal, EnteSpacing.xl)
                .padding(.vertical, EnteSpacing.md)
                .frame(minHeight: 44)
        }
        .buttonStyle(.plain)
        .foregroundStyle(selected ? .white : (enabled ? palette.mutedText : palette.disabledText))
        .background(selected ? palette.primary : palette.surface)
        .clipShape(RoundedRectangle(cornerRadius: EnteRadius.large, style: .continuous))
        .accessibilityAddTraits(selected ? .isSelected : [])
    }
}
